ipcs/ipcs.1
.Dd December 2008
.Dt IPCS 1

.Sh NAME

.Nm ipcs
.Nd report XSI interprocess communication facilities status

.Sh SYNOPSIS

.Nm ipcs
.Ar [-qms][-a | -bcopt]

.Sh DESCRIPTION

    The ipcs utility shall write information about active interprocess
communication facilities.

    Without options, information shall be written in short format for message
queues, shared memory segments, and semaphore sets that are currently active
in the system. Otherwise, the information that is displayed is controlled by
the options specified.

.Sh OPTIONS

    The ipcs facility supports the Base Definitions volume of IEEE Std
1003.1-2001, Section 12.2, Utility Syntax Guidelines.

    The ipcs utility accepts the following options:

    -q
        Write information about active message queues.
    -m
        Write information about active shared memory segments.
    -s
        Write information about active semaphore sets.

    If -q, -m, or -s are specified, only information about those facilities
shall be written. If none of these three are specified, information about all
three shall be written subject to the following options:

    -a
        Use all print options. (This is a shorthand notation for -b, -c, -o,
-p, and -t.)
    -b
        Write information on maximum allowable size. (Maximum number of bytes
in messages on queue for message queues, size of segments for shared memory,
and number of semaphores in each set for semaphores.)
    -c
        Write creator's user name and group name; see below.
    -o
        Write information on outstanding usage. (Number of messages on queue
and total number of bytes in messages on queue for message queues, and number
of processes attached to shared memory segments.)
    -p
        Write process number information. (Process ID of the last process to
send a message and process ID of the last process to receive a message on
message queues, process ID of the creating process, and process ID of the
last process to attach or detach on shared memory segments.)
    -t
        Write time information. (Time of the last control operation that
changed the access permissions for all facilities, time of the last msgsnd()
and msgrcv() operations on message queues, time of the last shmat() and
shmdt() operations on shared memory, and time of the last semop() operation
on semaphores.)

.Sh OPERANDS

    None.

.Sh STDIN

    Not used.

.Sh INPUT FILES

        The group database

        The user database

.Sh ENVIRONMENT VARIABLES

    The following environment variables shall affect the execution of ipcs:

    LANG
        Provide a default value for the internationalization variables that
are unset or null. (See the Base Definitions volume of IEEE Std 1003.1-2001,
Section 8.2, Internationalization Variables for the precedence of
internationalization variables used to determine the values of locale
categories.)
    LC_ALL
        If set to a non-empty string value, override the values of all the
other internationalization variables.
    LC_CTYPE
        Determine the locale for the interpretation of sequences of bytes of
text data as characters (for example, single-byte as opposed to multi-byte
characters in arguments).
    LC_MESSAGES
        Determine the locale that should be used to affect the format and
contents of diagnostic messages written to standard error.
    NLSPATH
        Determine the location of message catalogs for the processing of
LC_MESSAGES .
    TZ
        Determine the timezone for the date and time strings written by ipcs.
If TZ is unset or null, an unspecified default timezone shall be used.

.Sh ASYNCHRONOUS EVENTS

    Default.

.Sh STDOUT

    An introductory line shall be written with the format:

    "IPC status from %s as of %s\n", <source>, <date>

    where <source> indicates the source used to gather the statistics and
<date> is the information that would be produced by the date command when
invoked in the POSIX locale.

    The ipcs utility then shall create up to three reports depending upon the
-q, -m, and -s options. The first report shall indicate the status of message
queues, the second report shall indicate the status of shared memory
segments, and the third report shall indicate the status of semaphore sets.

    If the corresponding facility is not installed or has not been used since
the last reboot, then the report shall be written out in the format:

    "%s facility not in system.\n", <facility>

    where <facility> is Message Queue, Shared Memory, or Semaphore, as
appropriate. If the facility has been installed and has been used since the
last reboot, column headings separated by one or more spaces and followed by
a <newline> shall be written as indicated below followed by the facility name
written out using the format:

    "%s:\n", <facility>

    where <facility> is Message Queues, Shared Memory, or Semaphores, as
appropriate. On the second and third reports the column headings need not be
written if the last column headings written already provide column headings
for all information in that report.

    The column headings provided in the first column below and the meaning of
the information in those columns shall be given in order below; the letters
in parentheses indicate the options that shall cause the corresponding column
to appear; "all" means that the column shall always appear. Each column is
separated by one or more <space>s. Note that these options only determine
what information is provided for each report; they do not determine which
reports are written.

    T (all)
        Type of facility:

        q
            Message queue.
        m
            Shared memory segment.
        s
            Semaphore.

        This field is a single character written using the format %c.
    ID (all)
        The identifier for the facility entry. This field shall be written
using the format %d.
    KEY (all)
        The key used as an argument to msgget(), semget(), or shmget() to
create the facility entry.

        Note:
            The key of a shared memory segment is changed to IPC_PRIVATE when
the segment has been removed until all processes attached to the segment
detach it.

        This field shall be written using the format 0x%x.
    MODE (all)
        The facility access modes and flags. The mode shall consist of 11
characters that are interpreted as follows.

        The first character shall be:

        S
            If a process is waiting on a msgsnd() operation.
        -
            If the above is not true.

        The second character shall be:

        R
            If a process is waiting on a msgrcv() operation.
        C or -
            If the associated shared memory segment is to be cleared when the
first attach operation is executed.
        -
            If none of the above is true.

        The next nine characters shall be interpreted as three sets of three
bits each. The first set refers to the owner's permissions; the next to
permissions of others in the usergroup of the facility entry; and the last to
all others. Within each set, the first character indicates permission to
read, the second character indicates permission to write or alter the
facility entry, and the last character is a minus sign ( '-' ).

        The permissions shall be indicated as follows:

        r
            If read permission is granted.
        w
            If write permission is granted.
        a
            If alter permission is granted.
        -
            If the indicated permission is not granted.

        The first character following the permissions specifies if there is
an alternate or additional access control method associated with the
facility. If there is no alternate or additional access control method
associated with the facility, a single <space> shall be written; otherwise,
another printable character is written.
    OWNER (all)
        The user name of the owner of the facility entry. If the user name of
the owner is found in the user database, at least the first eight column
positions of the name shall be written using the format %s. Otherwise, the
user ID of the owner shall be written using the format %d.
    GROUP (all)
        The group name of the owner of the facility entry. If the group name
of the owner is found in the group database, at least the first eight column
positions of the name shall be written using the format %s. Otherwise, the
group ID of the owner shall be written using the format %d.

    The following nine columns shall be only written out for message queues:

    CREATOR (a,c)
        The user name of the creator of the facility entry. If the user name
of the creator is found in the user database, at least the first eight column
positions of the name shall be written using the format %s. Otherwise, the
user ID of the creator shall be written using the format %d.
    CGROUP (a,c)
        The group name of the creator of the facility entry. If the group
name of the creator is found in the group database, at least the first eight
column positions of the name shall be written using the format %s. Otherwise,
the group ID of the creator shall be written using the format %d.
    CBYTES (a,o)
        The number of bytes in messages currently outstanding on the
associated message queue. This field shall be written using the format %d.
    QNUM (a,o)
        The number of messages currently outstanding on the associated
message queue. This field shall be written using the format %d.
    QBYTES (a,b)
        The maximum number of bytes allowed in messages outstanding on the
associated message queue. This field shall be written using the format %d.
    LSPID (a,p)
        The process ID of the last process to send a message to the
associated queue. This field shall be written using the format:

        "%d", <pid>

        where <pid> is 0 if no message has been sent to the corresponding
message queue; otherwise, <pid> shall be the process ID of the last process
to send a message to the queue.
    LRPID (a,p)
        The process ID of the last process to receive a message from the
associated queue. This field shall be written using the format:

        "%d", <pid>

        where <pid> is 0 if no message has been received from the
corresponding message queue; otherwise, <pid> shall be the process ID of the
last process to receive a message from the queue.
    STIME (a,t)
        The time the last message was sent to the associated queue. If a
message has been sent to the corresponding message queue, the hour, minute,
and second of the last time a message was sent to the queue shall be written
using the format %d : %2.2d : %2.2d. Otherwise, the format " no-entry" shall
be written.
    RTIME (a,t)
        The time the last message was received from the associated queue. If
a message has been received from the corresponding message queue, the hour,
minute, and second of the last time a message was received from the queue
shall be written using the format %d : %2.2d : %2.2d. Otherwise, the format "
no-entry" shall be written.

    The following eight columns shall be only written out for shared memory
segments.

    CREATOR (a,c)
        The user of the creator of the facility entry. If the user name of
the creator is found in the user database, at least the first eight column
positions of the name shall be written using the format %s. Otherwise, the
user ID of the creator shall be written using the format %d.
    CGROUP (a,c)
        The group name of the creator of the facility entry. If the group
name of the creator is found in the group database, at least the first eight
column positions of the name shall be written using the format %s. Otherwise,
the group ID of the creator shall be written using the format %d.
    NATTCH (a,o)
        The number of processes attached to the associated shared memory
segment. This field shall be written using the format %d.
    SEGSZ (a,b)
        The size of the associated shared memory segment. This field shall be
written using the format %d.
    CPID (a,p)
        The process ID of the creator of the shared memory entry. This field
shall be written using the format %d.
    LPID (a,p)
        The process ID of the last process to attach or detach the shared
memory segment. This field shall be written using the format:

        "%d", <pid>

        where <pid> is 0 if no process has attached the corresponding shared
memory segment; otherwise, <pid> shall be the process ID of the last process
to attach or detach the segment.
    ATIME (a,t)
        The time the last attach on the associated shared memory segment was
completed. If the corresponding shared memory segment has ever been attached,
the hour, minute, and second of the last time the segment was attached shall
be written using the format %d : %2.2d : %2.2d. Otherwise, the format "
no-entry" shall be written.
    DTIME (a,t)
        The time the last detach on the associated shared memory segment was
completed. If the corresponding shared memory segment has ever been detached,
the hour, minute, and second of the last time the segment was detached shall
be written using the format %d : %2.2d : %2.2d. Otherwise, the format "
no-entry" shall be written.

    The following four columns shall be only written out for semaphore sets:

    CREATOR (a,c)
        The user of the creator of the facility entry. If the user name of
the creator is found in the user database, at least the first eight column
positions of the name shall be written using the format %s. Otherwise, the
user ID of the creator shall be written using the format %d.
    CGROUP (a,c)
        The group name of the creator of the facility entry. If the group
name of the creator is found in the group database, at least the first eight
column positions of the name shall be written using the format %s. Otherwise,
the group ID of the creator shall be written using the format %d.
    NSEMS (a,b)
        The number of semaphores in the set associated with the semaphore
entry. This field shall be written using the format %d.
    OTIME (a,t)
        The time the last semaphore operation on the set associated with the
semaphore entry was completed. If a semaphore operation has ever been
performed on the corresponding semaphore set, the hour, minute, and second of
the last semaphore operation on the semaphore set shall be written using the
format %d : %2.2d : %2.2d. Otherwise, the format " no-entry" shall be
written.

    The following column shall be written for all three reports when it is
requested:

    CTIME (a,t)
        The time the associated entry was created or changed. The hour,
minute, and second of the time when the associated entry was created shall be
written using the format %d : %2.2d : %2.2d.

.Sh STDERR

    The standard error shall be used only for diagnostic messages.

.Sh OUTPUT FILES

    None.

.Sh EXTENDED DESCRIPTION

    None.

.Sh EXIT STATUS

    The following exit values shall be returned:

     0
        Successful completion.
    >0
        An error occurred.

.Sh CONSEQUENCES OF ERRORS

    Default.

The following sections are informative.
.Sh APPLICATION USAGE

    Things can change while ipcs is running; the information it gives is
guaranteed to be accurate only when it was retrieved.

.Sh EXAMPLES

    None.

.Sh RATIONALE

    None.

.Sh FUTURE DIRECTIONS

    None.

.Sh SEE ALSO

    The System Interfaces volume of IEEE Std 1003.1-2001, msgrcv(), msgsnd(),
semget(), semop(), shmat(), shmdt(), shmget()



NEWS

Version 5 - unreleased

	- New utilities: cmp, tput, tabs, stty, uname, uniq

	- Bug fixes.

Version 4 - December 9, 2006

	- import codebase into git repository on git.kernel.org
	- fix compiler warnings on newer compilers
	- add x86-64 support
	- add WIP utility: pax
	- add utilities: env, id, strings, time
	- add simple "it works" tests
	- create per-util status files
	- drop hand-rolled 'compress', import BSD 'compress' instead

Version 3 - February 7, 2005

	- Add copyright/license info to each source file.
	- Make sure bug report address points to SourceForge.

	- New utilities: grep, pathchk, pwd, split, tty.
	- New work-in-progress (WIP) utility: mv.

	- Fix many 'compress' bugs.
	- Security fixes.

	- (internal) Move compress engine into lib.

Version 2 - January 26, 2005

	- Update README with project goals/manifesto,
	  and some URLs.

	- New utilities: compress, comm, date, what, wc, renice,
	  ipcrm, kill, rm.

	- Bug fixes.

Version 1 - August 21, 2004

	- Initial release.



ipcs/ipcs.cc

/*
 * Copyright 2008-2009 Red Hat, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#ifndef HAVE_CONFIG_H
#error missing autoconf-generated config.h.
#endif
#include "posixutils-config.h"

#ifndef _GNU_SOURCE
#define _GNU_SOURCE		/* for MSG_INFO, MSG_STAT */
#endif

#include <sys/types.h>
#include <sys/utsname.h>
#include <stdlib.h>
#include <string.h>
#include <argp.h>
#include <time.h>
#include <sys/ipc.h>
#include <sys/msg.h>
#include <sys/sem.h>
#include <sys/shm.h>
#include <libpu.h>


static const char doc[] =
N_("ipcs - report XSI interprocess communication facilities status");

static struct argp_option options[] = {
	{ "msg", 'q', NULL, 0,
	  N_("Write information about active message queues.") },
	{ "shm", 'm', NULL, 0,
	  N_("Write information about active shared memory segments.") },
	{ "sem", 's', NULL, 0,
	  N_("Write information about active semaphore sets.") },

	{ "all", 'a', NULL, 0,
	  N_("Use all print options. (This is a shorthand notation for -b, -c, -o, -p, and -t.)") },
	{ "max-size", 'b', NULL, 0,
	  N_("Write information on maximum allowable size.") },
	{ "creator", 'c', NULL, 0,
	  N_("Write creator's user name and group name") },
	{ "outstanding", 'o', NULL, 0,
	  N_("Write information on outstanding usage.") },
	{ "process", 'p', NULL, 0,
	  N_("Write process number information.") },
	{ "time", 't', NULL, 0,
	  N_("Write time information.") },

	{ }
};

/* X/Open says we must define this ourselves? */
union semun {
	int			val;
	struct semid_ds		*buf;
	unsigned short		*array;
	struct seminfo		*__buf;
};

static error_t parse_opt (int key, char *arg, struct argp_state *state);
static const struct argp argp = { options, parse_opt, NULL, doc };

enum parse_options_ibits {
	IOPT_MSG		= (1 << 0),
	IOPT_SHM		= (1 << 1),
	IOPT_SEM		= (1 << 2),
	IOPT_ALL		= IOPT_MSG | IOPT_SHM | IOPT_SEM,
};

enum parse_options_rbits {
	ROPT_SIZE		= (1 << 0),
	ROPT_CREATOR		= (1 << 1),
	ROPT_OUTST		= (1 << 2),
	ROPT_PROC		= (1 << 3),
	ROPT_TIME		= (1 << 4),

	ROPT_ALL		= ROPT_SIZE | ROPT_CREATOR | ROPT_OUTST |
				  ROPT_PROC | ROPT_TIME,
};

static unsigned int opt_info, opt_print;

static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	switch (key) {
	case 'q': opt_info |= IOPT_MSG; break;
	case 'm': opt_info |= IOPT_SHM; break;
	case 's': opt_info |= IOPT_SEM; break;

	case 'a': opt_print |= ROPT_ALL; break;
	case 'b': opt_print |= ROPT_SIZE; break;
	case 'c': opt_print |= ROPT_CREATOR; break;
	case 'o': opt_print |= ROPT_OUTST; break;
	case 'p': opt_print |= ROPT_PROC; break;
	case 't': opt_print |= ROPT_TIME; break;

	default:
		return ARGP_ERR_UNKNOWN;
	}

	return 0;
}

static int print_header(void)
{
	char timestr[64];
	time_t time_n;
	struct tm tm;
	struct utsname uts;

	if (uname(&uts) < 0) {
		perror("uname");
		return 1;
	}

	time_n = time(NULL);
	if (time_n == ((time_t)-1)) {
		perror("time");
		return 1;
	}

	localtime_r(&time_n, &tm);

	if (strftime(timestr, sizeof(timestr),
		     "%a, %d %b %Y %H:%M:%S %Z", &tm) < 15) {
		fprintf(stderr, _("stftime(3) failed\n"));
		return 1;
	}

	printf(_("IPC status from %s as of %s\n"), uts.nodename, timestr);

	return 0;
}

static int print_msg(void)
{
	int max_q, mq, rc;
	struct msginfo mi;

	printf(_("Message queues:\n"));

	max_q = msgctl(0, MSG_INFO, (struct msqid_ds *) &mi);
	if (max_q < 0) {
		perror("msgctl");
		return 1;
	}

	for (mq = 0; mq <= max_q; mq++) {
		struct msqid_ds ds;

		rc = msgctl(mq, MSG_STAT, &ds);
		if (rc < 0)
			continue;

		/* COLUMN: type of facility */
		/* COLUMN: identifier for this facility */
		/* COLUMN: key used as arg to create facility */
		/* COLUMN: The facility access modes and flags. */
		/* COLUMN: The user name/id of the owner of the entry */
		/* COLUMN: The group name/id of the owner of the entry */
		printf("m %d 0x%x --%c%c%c%c%c%c%c%c%c %d %d",
		       mq,
		       ds.msg_perm.__key,
		       ds.msg_perm.mode & 0400,	/* user read */
		       ds.msg_perm.mode & 0200,	/* user write */
		       ds.msg_perm.mode & 0200,	/* user alter */
		       ds.msg_perm.mode & 0040,	/* group read */
		       ds.msg_perm.mode & 0020,	/* group write */
		       ds.msg_perm.mode & 0020,	/* group alter */
		       ds.msg_perm.mode & 0004,	/* other read */
		       ds.msg_perm.mode & 0002,	/* other write */
		       ds.msg_perm.mode & 0002,	/* other alter */
		       ds.msg_perm.uid,
		       ds.msg_perm.gid);

		/* COLUMN: The user name of the creator of the entry */
		/* COLUMN: The group name of the creator of the entry */
		if (opt_print & ROPT_CREATOR)
			printf(" %d %d",
			       ds.msg_perm.cuid,
			       ds.msg_perm.cgid);

		/* COLUMN: number of bytes in messages currently outstanding */
		/* COLUMN: number of messages currently outstanding */
		if (opt_print & ROPT_OUTST)
			printf(" %lu %lu",
			       ds.__msg_cbytes,
			       ds.msg_qnum);

		if (opt_print & ROPT_SIZE)
			printf(" %lu", ds.msg_qbytes);

		if (opt_print & ROPT_PROC)
			printf(" %d %d",
			       ds.msg_lspid,
			       ds.msg_lrpid);

		/* FIXME: should be HHHHH:MM:SS */
		if (opt_print & ROPT_TIME)
			printf(" %lu %lu",
			       ds.msg_stime,
			       ds.msg_rtime);

		printf(" %lu\n", ds.msg_ctime);
	}

	return 0;
}

static int print_shm(void)
{
	int max_shm, shmid, rc;
	struct shminfo shi;

	printf(_("Shared memory:\n"));

	max_shm = shmctl(0, SHM_INFO, (struct shmid_ds *) &shi);
	if (max_shm < 0) {
		perror("shmctl");
		return 1;
	}

	for (shmid = 0; shmid <= max_shm; shmid++) {
		struct shmid_ds ds;

		rc = shmctl(shmid, SHM_STAT, &ds);
		if (rc < 0)
			continue;

		/* COLUMN: type of facility */
		/* COLUMN: identifier for this facility */
		/* COLUMN: key used as arg to create facility */
		/* COLUMN: The facility access modes and flags. */
		/* COLUMN: The user name/id of the owner of the entry */
		/* COLUMN: The group name/id of the owner of the entry */
		printf("m %d 0x%x --%c%c%c%c%c%c%c%c%c %d %d",
		       shmid,
		       ds.shm_perm.__key,
		       ds.shm_perm.mode & 0400,	/* user read */
		       ds.shm_perm.mode & 0200,	/* user write */
		       ds.shm_perm.mode & 0200,	/* user alter */
		       ds.shm_perm.mode & 0040,	/* group read */
		       ds.shm_perm.mode & 0020,	/* group write */
		       ds.shm_perm.mode & 0020,	/* group alter */
		       ds.shm_perm.mode & 0004,	/* other read */
		       ds.shm_perm.mode & 0002,	/* other write */
		       ds.shm_perm.mode & 0002,	/* other alter */
		       ds.shm_perm.uid,
		       ds.shm_perm.gid);

		/* COLUMN: The user name of the creator of the entry */
		/* COLUMN: The group name of the creator of the entry */
		if (opt_print & ROPT_CREATOR)
			printf(" %d %d",
			       ds.shm_perm.cuid,
			       ds.shm_perm.cgid);

		/* COLUMN: The number of processes attached to the segment */
		if (opt_print & ROPT_OUTST)
			printf(" %lu", ds.shm_nattch);

		if (opt_print & ROPT_SIZE)
			printf(" %lu", ds.shm_segsz);

		if (opt_print & ROPT_PROC)
			printf(" %d %d",
			       ds.shm_cpid,
			       ds.shm_lpid);

		/* FIXME: should be HHHHH:MM:SS */
		if (opt_print & ROPT_TIME)
			printf(" %lu %lu",
			       ds.shm_atime,
			       ds.shm_dtime);

		printf(" %lu\n", ds.shm_ctime);
	}

	return 0;
}

static int print_sem(void)
{
	int max_sem, semid, rc;
	struct seminfo sei;
	union semun arg;

	printf(_("Semaphores:\n"));

	arg.array = (unsigned short *) &sei;
	max_sem = semctl(0, 0, SEM_INFO, arg);
	if (max_sem < 0) {
		perror("semctl");
		return 1;
	}

	for (semid = 0; semid <= max_sem; semid++) {
		struct semid_ds ds;

		arg.buf = &ds;
		rc = semctl(semid, 0, SEM_STAT, arg);
		if (rc < 0)
			continue;

		/* COLUMN: type of facility */
		/* COLUMN: identifier for this facility */
		/* COLUMN: key used as arg to create facility */
		/* COLUMN: The facility access modes and flags. */
		/* COLUMN: The user name/id of the owner of the entry */
		/* COLUMN: The group name/id of the owner of the entry */
		printf("m %d 0x%x --%c%c%c%c%c%c%c%c%c %d %d",
		       semid,
		       ds.sem_perm.__key,
		       ds.sem_perm.mode & 0400,	/* user read */
		       ds.sem_perm.mode & 0200,	/* user write */
		       ds.sem_perm.mode & 0200,	/* user alter */
		       ds.sem_perm.mode & 0040,	/* group read */
		       ds.sem_perm.mode & 0020,	/* group write */
		       ds.sem_perm.mode & 0020,	/* group alter */
		       ds.sem_perm.mode & 0004,	/* other read */
		       ds.sem_perm.mode & 0002,	/* other write */
		       ds.sem_perm.mode & 0002,	/* other alter */
		       ds.sem_perm.uid,
		       ds.sem_perm.gid);

		/* COLUMN: The user name of the creator of the entry */
		/* COLUMN: The group name of the creator of the entry */
		if (opt_print & ROPT_CREATOR)
			printf(" %d %d",
			       ds.sem_perm.cuid,
			       ds.sem_perm.cgid);

		if (opt_print & ROPT_SIZE)
			printf(" %lu", ds.sem_nsems);

		/* FIXME: should be HHHHH:MM:SS */
		if (opt_print & ROPT_TIME)
			printf(" %lu",
			       ds.sem_otime);

		printf(" %lu\n", ds.sem_ctime);
	}

	return 0;
}

int main (int argc, char *argv[])
{
	pu_init();

	error_t argp_rc = argp_parse(&argp, argc, argv, 0, NULL, NULL);
	if (argp_rc) {
		fprintf(stderr, _("%s: argp_parse failed: %s\n"),
			argv[0], strerror(argp_rc));
		return EXIT_FAILURE;
	}

	if (!opt_info)
		opt_info = IOPT_ALL;
	if (!opt_print)
		opt_print = ROPT_PROC;

	if (print_header())
		return 1;

	if (opt_info & IOPT_MSG && print_msg())
		return 1;
	if (opt_info & IOPT_SHM && print_shm())
		return 1;
	if (opt_info & IOPT_SEM && print_sem())
		return 1;

	return 0;
}

